pkg/util/lsof/lsof_linux.go

```diff
@@ -0,0 +1,383 @@
+		// which aren't parsed by gopsutil, so we just assume it's memory mapped
+		fd = "mem"
+	} else if ty == "DIR" {
+		if path == "/" {
+			fd = "rtd"
+		} else if cwd != "" && path == cwd {
+			fd = "cwd"
+		}
+	}
+	return fd
+}
+
+func (ofl *openFilesLister) fdMetadata() (Files, error) {
+	openFDs, err := ofl.proc.FileDescriptors()
+	if err != nil {
+		return nil, err
+	}
+
+	var files Files
+	for _, openFD := range openFDs {
+		file, ok := ofl.fdStat(openFD)
+		if ok {
+			files = append(files, file)
+		}
+	}
+
+	return files, nil
+}
+
+func (ofl *openFilesLister) fdStat(fd uintptr) (File, bool) {
+	var err error
+	file := File{
+		Fd: fmt.Sprintf("%d", fd),
+	}
+
+	fdLinkPath := fmt.Sprintf("%s/fd/%d", ofl.procPIDPath(), fd)
+
+	var inode uint64
+	if file.Type, file.OpenPerm, _, _ = fileStats(ofl.lstat, fdLinkPath); file.Type == "" {
+		return File{}, false
+	}
+
+	var ok bool
+	// remove some unnecessary information from permissions string
+	// if the expectations are not met, the permissions are left as is
+
+	// file descriptors always have no sticky bit, setuid, setgid
+	file.OpenPerm = strings.TrimPrefix(file.OpenPerm, "-")
+	// file descriptors always have no permission for group and others
+	file.OpenPerm, ok = strings.CutSuffix(file.OpenPerm, "------")
+	if ok {
+		// file descriptors always have execute permission
+		file.OpenPerm = strings.TrimSuffix(file.OpenPerm, "x")
+	}
+
+	if file.Type, file.FilePerm, file.Size, inode = fileStats(ofl.stat, fdLinkPath); file.Type == "" {
+		return File{}, false
+	}
+
+	if info, ok := ofl.socketInfo[inode]; ok {
+		file.Name = info.Description
+		file.FilePerm = info.State
+		file.Type = info.Protocol
+	} else {
+		file.Name, err = ofl.readlink(fdLinkPath)
+		if err != nil {
+			return File{}, false
+		}
+	}
+
+	return file, true
+}
+
+// TCP state codes, from the Linux kernel
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/net/tcp_states.h
+//
+// The same codes are used for UDP, but only ESTABLISHED and CLOSE are valid
+var tcpStates = map[uint64]string{
+	1:  "ESTABLISHED",
+	2:  "SYN_SENT",
+	3:  "SYN_RECV",
+	4:  "FIN_WAIT1",
+	5:  "FIN_WAIT2",
+	6:  "TIME_WAIT",
+	7:  "CLOSE",
+	8:  "CLOSE_WAIT",
+	9:  "LAST_ACK",
+	10: "LISTEN",
+	11: "CLOSING",
+	12: "NEW_SYN_RECV",
+	13: "BOUND_INACTIVE",
+}
+
+func stateStr(state uint64) string {
+	if s, ok := tcpStates[state]; ok {
+		return s
+	}
+	return fmt.Sprintf("UNKNOWN(%d)", state)
+}
+
+// readSocketInfo reads the socket information from /proc/<pid>/net/{tcp,tcp6,udp,udp6,unix}
+// returns a map of inode to socketInfo
+// see https://www.kernel.org/doc/Documentation/networking/proc_net_tcp.txt
+func readSocketInfo(procPIDPath string) map[uint64]socketInfo {
+	si := make(map[uint64]socketInfo)
+
+	fs, err := procfs.NewFS(procPIDPath)
+	if err != nil {
+		log.Debugf("Failed to read %s: %s", procPIDPath, err)
+		return si
+	}
+
+	for protocol, parser := range map[string]func() (procfs.NetTCP, error){
```

Comment:
Why separating tcp and udp related protocol in different ranges? 

Could we group them all in a single one? 

```go
for protocol, parser := range map[string]func() (procfs.NetTCP, error){
    "tcp":  fs.NetTCP,
    "tcp6": fs.NetTCP6,
    "udp":  fs.NetUDP,
    "udp6": fs.NetUDP6,
}
```

Reply:
> I could use reflect to access the fields, but I thought it would just be less clear than a little bit of code duplication

Not worth it 😄 . Thanks for explaining the reason why we couldn't aggregate the all in a single range.